### PR TITLE
fix: eliminate branching from `TokenRouter`

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

Fixes bug in `TokenRouter` which dispatches messages to token recipient rather than address in `Router` table when hook overrides are provided.

### Drive-by changes

Simplifies inheritance tree of `TokenRouter` > `GasRouter` > `Router` > `MailboxClient`

### Backward compatibility

Yes

### Testing

Unit Tests

### CODE REVIEW

File: solidity/contracts/client/Router.sol [hunk 1]
```diff
@@ -167,28 +167,67 @@ abstract contract Router is MailboxClient, IMessageRecipient {
...
+            mailbox.dispatch{value: _value}(
+                _destinationDomain,
+                _router,
+                _messageBody,
+                _hookMetadata,
+                IPostDispatchHook(_hook)
+            );
     }
 
+    // for backward compatibility with v2 contracts
```
Comment (nambrot) [inline solidity/contracts/client/Router.sol:188]: Why is this necessary?

File: solidity/contracts/token/libs/TokenRouter.sol [hunk 2]
```diff
@@ -97,45 +90,45 @@ abstract contract TokenRouter is GasRouter {
...
         bytes32 _recipient,
         uint256 _amountOrId,
-        uint256 _gasPayment,
-        bytes memory _hookMetadata,
-        address _hook
+        uint256 _value
     ) internal returns (bytes32 messageId) {
-        bytes memory metadata = _transferFromSender(_amountOrId);
-
-        if (address(_hook) == address(0)) {
```
Comment (nambrot) [inline solidity/contracts/token/libs/TokenRouter.sol:123]: Is it possible to add a test case for this?

**Review by nambrot — COMMENTED** [review]
Please add a description of the bug that this unearthed

**Review by nbayindirli — APPROVED** [review]
lgtm

**Review by nambrot — APPROVED** [review]
